rust_src/src/multibyte.rs

```diff
@@ -348,13 +357,214 @@ impl LispStringRef {
         s.size
     }
 
-    /// Return width of STRING when displayed in the current buffer. Width is
-    /// measured by how many columns it occupies on the screen. When calculating
-    /// width of a multibyte character in STRING, only the base leading-code is
-    /// considered; the validity of the following bytes is not checked.  Tabs in
-    /// STRING are always taken to occupy `tab-width' columns.
+    /// Return the byte index corresponding to the given character index.
+    ///
+    /// Caches the ouput of the most recent calculation to make several calls on
+    /// the same string faster.
+    pub fn char_idx_to_byte(self, char_index: usize) -> usize {
+        let mut best_below = 0;
+        let mut best_below_byte = 0;
+        let mut best_above = self.len_chars() as usize;
+        let mut best_above_byte = self.len_bytes() as usize;
+
+        // Return if contents are unibyte.
+        if best_above == best_above_byte {
+            return char_index;
+        }
+
+        // Use cache operating on the cached string?
+        unsafe {
+            if LispObject::from(self) == string_index_cache {
```

Comment:
Why store the LispObject instead of a `Option<LispStringRef>`?

Reply:
I only realized this now, but from what I can tell storing a `LispObject` is necessary for making `staticpro` work, as the pointer supplied to it has to point to a static variable, and storing an `ExternalPtr` but supplying a `*mut LispObject` to `staticpro` would require converting to a tagged pointer/LispObject at the time of invocation. This would mean that the variable only lasts the length of the funtion call. This is a by-product of a pointer to a LispObject essentially being two layers of pointers.